CFT-3212 - Feature/async

kudj opened the pull request:
## Add async HTTP client support

This PR adds an async variant of the HTTP client alongside the existing synchronous implementation.

### What's new

- **AsyncHttpClient**: New async client built on `httpx` with the same interface as the sync client
  - Async/await support for non-blocking I/O operations
  - Built-in rate limiting via `aiolimiter`
  - Same retry logic, error handling, and parameter merging as the sync client

- **Test coverage**: Comprehensive async test suite with proper mock handling

- **Documentation**: Updated README with async examples and usage patterns

- **Modernization**: 
  - Migrated from `setup.py` to `pyproject.toml`
  - Added `uv` for dependency management
  - Updated type hints to use modern syntax (`dict | None` instead of `Optional[Dict]`)
  - Cleaned up imports and removed deprecated patterns

### Breaking changes

None - the sync `HttpClient` API remains unchanged.

SgtMarmite committed:
rename http.py to client.py due to causing No Module Found error for http.cookiejar

SgtMarmite committed:
httpx initial implementation

SgtMarmite committed:
methods update, post, patch, delete

SgtMarmite committed:
update methods

SgtMarmite committed:
httpx updates

SgtMarmite committed:
add raw methods

SgtMarmite committed:
update client reference

SgtMarmite committed:
cleanup, readme update

SgtMarmite committed:
move examples to separate folder

SgtMarmite committed:
add new tests

SgtMarmite committed:
update tests

SgtMarmite committed:
finalize tests

SgtMarmite committed:
add Storage API example to docs

SgtMarmite committed:
update backoff calculation

SgtMarmite committed:
remove unused imports/installs

SgtMarmite committed:
remove more unused imports/installs

SgtMarmite committed:
add update_auth_header method + tests

SgtMarmite committed:
Update async_storage_client.py

SgtMarmite committed:
add correct typehints

SgtMarmite committed:
enable rate limiting using aiolimiter

SgtMarmite committed:
add process_multiple method and example

SgtMarmite committed:
refactor retry strategy

SgtMarmite committed:
fix tests

SgtMarmite committed:
add workaround to disable INFO msgs coming from httpx library

SgtMarmite committed:
add retry warning message

SgtMarmite committed:
reduce debug messages

SgtMarmite committed:
remove debug param

SgtMarmite committed:
revert logging setup

SgtMarmite committed:
silence httpcore debug messages

Commit:
reenable debug

Commit:
Update async_client.py

Commit:
Update async_client.py

Commit:
bump httpx version

Commit:
make max_requests per second float

kudj committed:
added detailed exception message

kudj committed:
added detailed exception message

kudj committed:
switch to python 3.12

kudj committed:
flake8

kudj committed:
fix message

kudj committed:
resolve conflicts

soustruh committed:
HTTPError has no attribute response

soustruh commented on .gitignore:
```diff
@@ -139,4 +139,7 @@ dmypy.json
+.DS_Store
+
+# Local test script
+/docs/examples/test.py
```
OK, this makes no sense, if anyone used their custom test script, it should be either added for everyone (considering it's in the docs/examples folder?) or just saved elsewhere or treated as a "WIP mess" during local development and ignored – but not in a `.gitignore` way. Creating test.py in the docs/examples folder in the future doesn't seem to be such an improbable scenario, in which case this line here would just cause confusion.

soustruh replied:
Updated

soustruh commented on README.md:
````diff
@@ -203,4 +203,74 @@ class KBCStorageClient(HttpClient):
+asyncio.run(main())
+```
+
+The AsyncHttpClient provides similar initialization and request methods as the HttpClient. 
````
Use "similar to" please, not "similar as"

soustruh replied:
Updated

soustruh commented on README.md:
```diff
@@ -203,4 +203,74 @@ class KBCStorageClient(HttpClient):
+The request methods return awaitable coroutines that can be awaited in an asynchronous context.
+
+#### Building HTTP client based on AsyncHttpClient Example
+This example demonstrates the default use of the HTTPClient as a base for REST API clients.
```
AsyncHttpClient

soustruh replied:
Updated

soustruh commented on README.md:
```diff
@@ -203,4 +203,74 @@ class KBCStorageClient(HttpClient):
+
+    async def get_files(self, show_expired=False):
+        params = {"showExpired": show_expired}
+        response = await self.get('tables', params=params, timeout=5)
```
I know this is just a docs example, but my eyes simply can't handle mixing quotes like this 🫣 🙈 😁

soustruh replied:
Updated

soustruh commented on docs/examples/pokeapi_async.py:
Typo in the file name 😇

soustruh replied:
Fixed

soustruh commented on docs/examples/pokeapi_process_multiple.py:
```diff
@@ -0,0 +1,44 @@
+import asyncio
+import csv
+import time
+from typing import List
```
Please don't use this deprecated class for typing. I also don't understand why this file is type annotated at ONE place, while the previous file has no annotations at all.

soustruh replied:
Updated 😌

soustruh commented on src/keboola/http_client/http.py:
```diff
@@ -6,7 +6,7 @@
 import requests
 from requests.adapters import HTTPAdapter
-from requests.packages.urllib3.util.retry import Retry
+from requests.packages.urllib3.util.retry import Retry  # noqa
```
Not needed at all 👀

soustruh replied:
OK, it WAS used, but since the import was put down incorrently (LOL # noqa 😏), my IDE didn't find the reference correctly. Fixed to:
```py
from urllib3.util import Retry
```
just as stated right in the requests docs: https://requests.readthedocs.io/en/latest/user/advanced/#example-automatic-retries

soustruh commented on src/keboola/http_client/async_client.py:
```diff
@@ -0,0 +1,241 @@
+
+                logging.error(f"Retry attempt {retry_attempt + 1} for {method} request to {url}: {message}")
+
+    async def get(self, endpoint: Optional[str] = None, **kwargs) -> Dict[str, Any]:
```
OK, I just checked the rest of this component, I have no words… I miss the basic motivations behind this renaming/remapping of industry standard Python packages 🤷‍♂️

soustruh replied:
Cannot fix this without breaking the components using this library 😿

soustruh commented on src/keboola/http_client/async_client.py:
```diff
@@ -0,0 +1,241 @@
+
+        self.limiter = None
+        if max_requests_per_second:
+            one_reqeust_per_second_amount = float(1 / max_requests_per_second)
```
Typo and misleading variable name

soustruh replied:
Fixed 😀

soustruh replied:
The funny thing is, instead of:
```py
        if max_requests_per_second:
            max_request_duration = float(1 / max_requests_per_second)
            self.limiter = AsyncLimiter(1, max_request_duration)
```
there could just be:
```py
        if max_requests_per_second:
            self.limiter = AsyncLimiter(max_requests_per_second, 1)
```

kudj committed:
retry on httpx.HTTPError, httpx.ReadError, httpx.ConnectError, httpx.ReadTimeout

kudj committed:
test

kudj committed:
Merge pull request #13 from keboola/feature/async-retry

Feature/async retry

soustruh committed:
simplified according to our actual needs

aligned with the latest changes in the cookiecutter template 🍪

soustruh committed:
minor wording changes

soustruh committed:
ruff formatting in code excerpts 🐶

soustruh committed:
filename typo ⚡️

soustruh committed:
removed deprecated type annotations (from typing import …) + ruff 🐶

soustruh committed:
fixed import (bye noqa) 👷‍♂️

soustruh committed:
misleading variable name

soustruh committed:
consistent headings

soustruh committed:
flake8.cfg consistent with cookiecutter template + minor formatting fixes 🍪

soustruh committed:
uv 💜

soustruh committed:
from __future__ import annotations for legacy Python versions

soustruh committed:
use the latest actions versions

soustruh committed:
Merge branch 'main' of github.com:keboola/python-http-client into feature/async

soustruh committed:
version replacement in CI

soustruh commented on pyproject.toml:
```diff
@@ -0,0 +1,50 @@
+[project]
+name = "keboola.http-client"
```
This is intentional, the package name at PyPI is dashed anyway.

soustruh commented on README.md:
```diff
@@ -117,93 +133,177 @@ else:
+### AsyncHttpClient
+
+The package also provides an asynchronous version of the HTTP client called AsyncHttpClient.
+It allows you to make asynchronous requests using async/await syntax. To use the AsyncHttpClient, import it from keboola.http_client_async:
```
Nice catch! 👍

soustruh commented on src/keboola/http_client/http.py:
```diff
@@ -1,17 +1,18 @@
+from urllib3.util import Retry
 
-Cookie = Union[Dict[str, str], CookieJar]
+Cookie = "dict[str, str] | CookieJar"
```
Unfortunately, this doesn't work in Python 3.8 which we still support with this component 😐

soustruh commented on src/keboola/http_client/http.py:
```diff
@@ -136,44 +145,43 @@ def _request_raw(self, method: str, endpoint_path: Optional[str] = None, **kwarg
 
         # Default parameters
         if self._default_params is not None:
             all_pars = {**params, **self._default_params}
```
Nice catch, but as this line wasn't modified in this PR, I won't touch this now… I will consider a bugfix in the near future though. 👀

soustruh commented on src/keboola/http_client/async_client.py:
```diff
@@ -0,0 +1,257 @@
+                    else:
+                        raise type(e)(error_msg) from e
+
+                backoff = self.backoff_factor**retry_attempt
```
Yeah, not to mention for factors lower than 1, the backoff is actually speeding up instead of backing off. 😁

soustruh commented on README.md:
```diff
@@ -117,93 +133,177 @@ else:
+            # Handle the error
+            pass
+
+
```
Yep, good catch. 👍 All this logic calls for simplification and cleanup anyway… Let's do this in another minor release after we finally release this. 🙃

soustruh commented on src/keboola/http_client/async_client.py:
```diff
@@ -0,0 +1,257 @@
+                    error_msg = f"Error '{message}' for url '{url}'"
+
+                if retry_attempt == self.retries:
+                    if isinstance(e, httpx.HTTPStatusError):
+                        raise
+                    else:
+                        raise type(e)(error_msg) from e
```
Not a problem, we just produce a little less logs for common, easily readable errors

soustruh commented on src/keboola/http_client/async_client.py:
```diff
@@ -0,0 +1,257 @@
+        is_absolute_path = kwargs.pop("is_absolute_path", False)
+        url = await self._build_url(endpoint, is_absolute_path)
+
+        all_params = {**self.default_params, **(params or {})}
```
Nice catch, this code wasn't part of the PR, but it was incorrect and too complicated anyway 🙄 👍

soustruh committed:
corrected async client docs

soustruh committed:
backoff wasn't implemented correctly

soustruh committed:
parameter precedence corrected & simplified

soustruh committed:
cleaning up log noise 🧹

The issue was that when mock_request didn't have a return value configured and returned a MagicMock object instead, so when response.raise_for_status() was called, the mock returned a coroutine that was never awaited.

soustruh committed:
use @patch.object decorator consistently

soustruh committed:
string literal typing comments 💬

ZdenekSrotyr reviewed (approved):
LGTM, tests look solid and everything reads well.

soustruh reviewed (approved):
LGTM, the updated client was thoroughly tested in many components already, the test coverage is also sufficient 👍

soustruh committed:
python 3.14 support + updated dependencies 🐍
